src/main/java/jcifs/smb1/smb1/Dfs.java

```diff
@@ -0,0 +1,328 @@
+        }
+        return null;
+    }
+    public boolean isTrustedDomain(String domain,
+                    NtlmPasswordAuthentication auth) throws SmbAuthException
+    {
+        HashMap domains = getTrustedDomains(auth);
+        if (domains == null)
+            return false;
+        domain = domain.toLowerCase();
+        return domains.get(domain) != null;
+    }
+    public SmbTransport getDc(String domain,
+                    NtlmPasswordAuthentication auth) throws SmbAuthException {
+        if (DISABLED)
+            return null;
+
+        try {
+            UniAddress addr = UniAddress.getByName(domain, true);
+            SmbTransport trans = SmbTransport.getSmbTransport(addr, 0);
+            DfsReferral dr = trans.getDfsReferrals(auth, "\\" + domain, 1);
+            if (dr != null) {
+                DfsReferral start = dr;
+                IOException e = null;
+
+                do {
+                    try {
+                        addr = UniAddress.getByName(dr.server);
+                        return SmbTransport.getSmbTransport(addr, 0);
+                    } catch (IOException ioe) {
+                        e = ioe;
+                    }
+
+                    dr = dr.next;
+                } while (dr != start);
+
+                throw e;
+            }
+        } catch (IOException ioe) {
+            if (log.level >= 3)
+                ioe.printStackTrace(log);
+            if (strictView && ioe instanceof SmbAuthException) {
+                throw (SmbAuthException)ioe;
+            }
+        }
+        return null;
+    }
+    public DfsReferral getReferral(SmbTransport trans,
+                    String domain,
+                    String root,
+                    String path,
+                    NtlmPasswordAuthentication auth) throws SmbAuthException {
+        if (DISABLED)
+            return null;
+
+        try {
+            String p = "\\" + domain + "\\" + root;
+            if (path != null)
+                p += path;
+            DfsReferral dr = trans.getDfsReferrals(auth, p, 0);
+            if (dr != null)
+                return dr;
+        } catch (IOException ioe) {
+            if (log.level >= 4)
+                ioe.printStackTrace(log);
+            if (strictView && ioe instanceof SmbAuthException) {
+                throw (SmbAuthException)ioe;
+            }
+        }
+        return null;
+    }
+    public synchronized DfsReferral resolve(String domain,
+                String root,
+                String path,
+                NtlmPasswordAuthentication auth) throws SmbAuthException {
+        DfsReferral dr = null;
+        long now = System.currentTimeMillis();
+
+        if (DISABLED || root.equals("IPC$")) {
+            return null;
+        }
+        /* domains that can contain DFS points to maps of roots for each
+         */
+        HashMap domains = getTrustedDomains(auth);
+        if (domains != null) {
+            domain = domain.toLowerCase();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;domain&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;path&#39;, &#39;root&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, reduce the method&#39;s visibility.

Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).